js/hw_2_js_functions_state/src/game_of_snake/game_of_snake.js

```diff
@@ -1,3 +1,409 @@
+import { emit } from "process";
 import { createEngine, ENGINE_INITIALIZE_SIGNAL } from "../engine/engine";
 
-// Good luck!
+
+const meshSize = 625;
+
+const initialState = {
+  snakeHead: [10, 10],
+  snakeBody: [
+    [10,10],
+    [10,11],
+    [10,12],
+    [10,13],
+    [10,14]
+  ],
+  food: [[15, 15]],
+  direction: 'up',
+  getWin: false,
+  getLose: false,
+  timer: 120000,
+  timeGone: 0,
+  score: 0,
+  pause: false
+}
+
+
+const engine = createEngine();
+
+engine.addSignalReducer(ENGINE_INITIALIZE_SIGNAL, () => {
+  return initialState;
+});
+
+engine.addSignalReducer("restartGame", () => {
+  return initialState
+});
+
+engine.addSignalReducer('setTimer',state => {
+  if(state.getWin || state.getLose  || state.pause) return state
+  
+  if(state.timer === 0){
+      return {
+          ...state,
+          getLose: true
+      }
+  }
+
+  return {
+      ...state,
+      timer: state.timer - 100,
+      timeGone: state.timeGone + 100
+  }
+})
+
+
+engine.addSignalReducer('up', state => {
+  if(state.getWin || state.getLose || state.pause) return state
+
+
+  if(state.direction === 'up' || state.direction === 'down'){
+    return state;
+  }
+  return {
+      ...state,
+      direction: 'up',
+  };
+})
+
+engine.addSignalReducer('down', state => {
+  if(state.getWin || state.getLose || state.pause) return state
+
+
+  if(state.direction === 'down'  || state.direction === 'up'){
+    return state;
+  }
+  return {
+      ...state,
+      direction: 'down',
+  };
+})
+
+engine.addSignalReducer('left', state => {
```

Comment:
"Left", "right", "top", and "down" reducers have much in common. Try to think how they can be implemented to avoid code duplication. Maybe you don't need 4 separate reducers?
```
emit('turn', 'right')

addSignalReducer('turn', (direction) => {...})
```